File: Node/core/lib/services/favorites-manager.js
"use strict";
var FavoritesManager = (function () {
    function FavoritesManager(userData) {
        this.userData = userData;
        this.maxFavoriteCount = 5;
        this.favoritesKey = 'favorites';
    }
    FavoritesManager.prototype.maxCapacityReached = function () {
        return this.getFavorites().length >= this.maxFavoriteCount;
    };
    FavoritesManager.prototype.isFavorite = function (location) {
        var favorites = this.getFavorites();
        for (var i = 0; i < favorites.length; i++) {
            if (this.areEqual(favorites[i].location, location)) {
                return true;
            }
        }
        return false;
    };
    FavoritesManager.prototype.isFavoriteLocationName = function (name) {
        var favorites = this.getFavorites();
        for (var i = 0; i < favorites.length; i++) {
            if (favorites[i].name.toLowerCase() === name.toLowerCase()) {
                return true;
            }
        }
        return false;
    };
    FavoritesManager.prototype.getFavoriteByIndex = function (index) {
        var favorites = this.getFavorites();
        if (index >= 0 && index < favorites.length) {
            return favorites[index];
        }
        return null;
    };
    FavoritesManager.prototype.getFavoriteByName = function (name) {
        var favorites = this.getFavorites();
        for (var i = 0; i < favorites.length; i++) {
            if (favorites[i].name.toLowerCase() === name.toLowerCase()) {
                return favorites[i];
            }
        }
        return null;
    };
    FavoritesManager.prototype.add = function (favoriteLocation) {
        var favorites = this.getFavorites();
        if (favorites.length >= this.maxFavoriteCount) {
            throw ('The max allowed number of favorite locations has already been reached.');
        }
        favorites.push(favoriteLocation);
        this.userData[this.favoritesKey] = favorites;
    };
    FavoritesManager.prototype.delete = function (favoriteLocation) {
        var favorites = this.getFavorites();
        var newFavorites = [];
        for (var i = 0; i < favorites.length; i++) {
            if (!this.areEqual(favorites[i].location, favoriteLocation.location)) {
                newFavorites.push(favorites[i]);
            }
        }
        this.userData[this.favoritesKey] = newFavorites;
    };
    FavoritesManager.prototype.update = function (currentValue, newValue) {
        var favorites = this.getFavorites();
        var newFavorites = [];
        for (var i = 0; i < favorites.length; i++) {
            if (this.areEqual(favorites[i].location, currentValue.location)) {
                newFavorites.push(newValue);
            }
            else {
                newFavorites.push(favorites[i]);
            }
        }
        this.userData[this.favoritesKey] = newFavorites;
    };
    FavoritesManager.prototype.getFavorites = function () {
        var storedFavorites = this.userData[this.favoritesKey];
        if (storedFavorites) {
            return storedFavorites;
        }
        else {
            return [];
        }
    };
    FavoritesManager.prototype.areEqual = function (location0, location1) {
        return location0.address.formattedAddress === location1.address.formattedAddress;
    };
    return FavoritesManager;
}());
exports.FavoritesManager = FavoritesManager;

File: CSharp/BotBuilderLocation/FavoritesManager.cs
namespace Microsoft.Bot.Builder.Location
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Bing;
    using Builder.Dialogs.Internals;
    using Dialogs;

    [Serializable]
    internal class FavoritesManager : IFavoritesManager
    {
        private const string FavoritesKey = "favorites";
        private const int MaxFavoriteCount = 5;

        public bool MaxCapacityReached(IBotData botData)
        {
            return this.GetFavorites(botData).Count >= MaxFavoriteCount;
        }

        public bool IsFavorite(IBotData botData, Location location)
        {
            var favorites = this.GetFavorites(botData);
            return favorites.Any(favoriteLocation => AreEqual(location, favoriteLocation.Location));
        }

        public bool IsFavoriteLocationName(IBotData botData, string name)
        {
            var favorites = this.GetFavorites(botData);
            return favorites.Any(favoriteLocation => StringComparer.OrdinalIgnoreCase.Equals(name, favoriteLocation.Name));
        }

        public FavoriteLocation GetFavoriteByIndex(IBotData botData, int index)
        {
            var favorites = this.GetFavorites(botData);

            if (index >= 0 && index <  favorites.Count)
            {
                return favorites[index];
            }

            return null;
        }

        public FavoriteLocation GetFavoriteByName(IBotData botData, string name)
        {
            var favorites = this.GetFavorites(botData);
            return favorites.Where(favoriteLocation => StringComparer.OrdinalIgnoreCase.Equals(name, favoriteLocation.Name)).FirstOrDefault();
        }

        public void Add(IBotData botData, FavoriteLocation value)
        {
            var favorites = this.GetFavorites(botData);

            if (favorites.Count >= MaxFavoriteCount)
            {
                throw new InvalidOperationException("The max allowed number of favorite locations has already been reached.");
            }

            favorites.Add(value);
            botData.UserData.SetValue(FavoritesKey, favorites);
        }

        public void Delete(IBotData botData, FavoriteLocation value)
        {
            var favorites = this.GetFavorites(botData);
            var newFavorites = new List<FavoriteLocation>();
            
            foreach (var favoriteItem in favorites)
            {
                if (!AreEqual(favoriteItem.Location, value.Location))
                {
                    newFavorites.Add(favoriteItem);
                }
            }

            botData.UserData.SetValue(FavoritesKey, newFavorites);
        }

        public void Update(IBotData botData, FavoriteLocation currentValue, FavoriteLocation newValue)
        {
            var favorites = this.GetFavorites(botData);
            var newFavorites = new List<FavoriteLocation>();

            foreach (var item in favorites)
            {
                if (AreEqual(item.Location, currentValue.Location))
                {
                    newFavorites.Add(newValue);
                }
                else
                {
                    newFavorites.Add(item);
                }
            }

            botData.UserData.SetValue(FavoritesKey, newFavorites);
        }

        public IList<FavoriteLocation> GetFavorites(IBotData botData)
        {
            List<FavoriteLocation> favorites;

            if (!botData.UserData.TryGetValue(FavoritesKey, out favorites))
            {
                // User currently has no favorite locations. Return an empty list.
                favorites = new List<FavoriteLocation>();
            }

            return favorites;
        }

        private static bool AreEqual(Location x, Location y)
        {
            // Other attributes of a location such as its Confidence, BoundaryBox, etc
            // should not be considered as distinguishing factors.
            // On the other hand, attributes of a location that are shown to the users
            // are what distinguishes one location from another. 
            return x.GetFormattedAddress(",") == y.GetFormattedAddress(",");
        }
    }
}
